### ISSUE TITLE
KafkaListener does not try to resolve type for ConsumerRecord

### ISSUE THREAD

**Issue by ST-DDT** [spring-projects/spring-kafka#1225]
🐞 Bug report / 🎁 Enhancement

According to your [docs](https://docs.spring.io/spring-kafka/reference/html/#kafka-listener-annotation)

I can create a `@KafkaListener` with a `ConsumerRecord<K, V>` like this:

````java
@KafkaListener(id = "thing2")
public void listen(ConsumerRecord<String, MyType> record) {
    ...
}
````

However in that case, the [`MessagingMessageListenerAdapter#determineInferredType(Method)`](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java#L477) method does not search for/resolve the `MyType` type and thus the `MessagingMessageConverter` is invoked with the fallback type `Object`.

(AFAICT this is defined in [`MessagingMessageListenerAdapter#eligibleParameter(MethodParameter)`](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java#L594))

**Callstack**

* org.springframework.kafka.support.converter.MessagingMessageConverter.extractAndConvertValue(ConsumerRecord<?, ?>, Type)
* org.springframework.kafka.support.converter.MessagingMessageConverter.toMessage(ConsumerRecord<?, ?>, Acknowledgment, Consumer<?, ?>, Type)
* org.springframework.kafka.listener.adapter.MessagingMessageListenerAdapter.toMessagingMessage(ConsumerRecord<K, V>, Acknowledgment, Consumer<?, ?>)

**TLDR**

The given `type` is just `Object` in [`RecordMessageConverter#toMessage(ConsumerRecord, Acknowledgment, Consumer, Type)`](https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/support/converter/RecordMessageConverter.java#L44) when the `@KafkaListener` uses a `ConsumerRecord<Key, Value>` parameter but it should be of type `Value`.

**Comment by garyrussell** [comment]
There is no conversion performed on ConsumerRecord, only POJO or Message<MyType> parameters. For a typed consumer record value you need an appropriate value deserializer so that Kake creates the right type.

**Comment by ST-DDT** [comment]
> There is no conversion performed on ConsumerRecord

Well... my custom RecordMessageConverter is invoked, so it at least it tries to do so.

**Comment by artembilan** [comment]
One more time: Spring can convert a `value` from the `ConsumerRecord` to put into method invocation arguments, but when you request a `ConsumerRecord` as a method arguments, there is no any Spring conversions.
Everything is up to native Apache Kafka deserialization. So, you need to configure an appropriate `value.deserializer` to get an expected type in the `ConsumerRecord` to propagate.
The point is that `ConsumerRecord` is created for us in the Apache Kafka client with already deserialized key and value. We just can't mutate such an object on the Spring level.

So, let's treat this as *Works as Designed* and close!

**Comment by garyrussell** [comment]
Just to be clear. Yes the converter is called to create a Message from the consumer record but the results are discarded because later up the stack we find the method wants the raw record. We might be able to skip the conversion but it might be tricky based on the layered architecture.

**Comment by ST-DDT** [comment]
> Just to be clear. Yes the converter is called to create a Message from the consumer record

This is what I meant. Due to the missing type information my converter fails, thus it either shouldn't be invoked or it should be provided with a hint such as `ConsumerRecord` that the value is supposed to be ignored. Using plain `Object` for that is bad since you cannot differentiate it from a valid value.

**Comment by garyrussell** [comment]
So don't configure the converter. The default converter will just use the record value as the payload of the discarded Message&lt;?&gt;.

**Comment by ST-DDT** [comment]
Well it autoconfigures itself by just being present in the application context. Maybe i should mention that i'm using spring boot.

**Comment by ST-DDT** [comment]
Well, if I do that then I just start pilling more and more workarounds.

1) Add two converters to the app context to avoid configuring it
2) Then configure some of the listeners to actually use the correct one
3) Then blow apart spring (boot) kafka in order to distinguish which listener actually needs to be configured
4) Drop spring kafka, because I implemented most of the logic myself anyway

I know this is probably a little bit exaggerated, but all I want to know is whether I'm supposed to convert it in my converter or not.

**Comment by garyrussell** [comment]
We will add some code to detect if conversion is not necessary and avoid it; I was just suggesting a work around until then.
